### ISSUE TITLE
"undefined is not an object" in coalesceChanges

### ISSUE THREAD

**Issue by tmcw** [mapbox/mapbox-gl-js#11784]
**mapbox-gl-js version**: 2.8.0

**browser**: safari 14.1.2, chrome 100, safari 15.4

### Steps to Trigger Behavior

 1. Use setFeatureState in combination with fairly quickly updating a style

### Link to Demonstration

TBD, this is relatively difficult to replicate in isolation

### Expected Behavior

setFeatureState should not cause an uncaught error

### Actual Behavior

A TypeError is produced:

```
TypeError: undefined is not an object (evaluating 'delete this.state[r][t][e]')
  at coalesceChanges(./node_modules/mapbox-gl/dist/mapbox-gl.js:31:362514)
  at prepare(./node_modules/mapbox-gl/dist/mapbox-gl.js:31:373744)
  at render(./node_modules/mapbox-gl/dist/mapbox-gl.js:35:310627)
  at _render(./node_modules/mapbox-gl/dist/mapbox-gl.js:35:447217)
  at apply(./node_modules/mapbox-gl/dist/mapbox-gl.js:35:452708)
  at o(./node_modules/@sentry/browser/esm/helpers.js:74:23)
```

**Comment by mourner** [comment]
@tmcw can you try adding a check like this in `source_state.js` and seeing if you can still reproduce this with a custom GL JS build?

```diff
-else {
+else if (this.state[sourceLayer][feature]) {
    for (const key of Object.keys(this.deletedStates[sourceLayer][feature])) {
        delete this.state[sourceLayer][feature][key];
    }
}

**Comment by tmcw** [comment]
Yes, sorry for the delay - I'll try making that mod and seeing if it's a fix.

**Comment by nick-romano** [comment]
Hello, I've been experiencing this same issue. 

Seems like the exact same spot

![image](https://user-images.githubusercontent.com/8332517/208514243-8543f490-9fca-4618-8dfa-f5d56541ef21.png)

Seems like a pretty straightforward fix, would you accept a PR?

**Comment by mourner** [comment]
> Seems like a pretty straightforward fix, would you accept a PR?

Definitely, thank you 🙏

**Comment by stepankuzmin** [comment]
Hi @enersis-pst,

This should be fixed in the latest beta [v2.13.0-beta.1](https://github.com/mapbox/mapbox-gl-js/releases/tag/v2.13.0-beta.1).
